Snap Sync: consensus: handle legacy pre-bedrock header verification

protolambda opened the pull request:
**Description**

Handle legacy pre-bedrock block-headers in consensus validation, to support snap-syncing without prior data-dir.

Very experimental. See optimism monorepo PR for context.

trianglesphere commented on consensus/beacon/consensus.go:
```diff
@@ -461,6 +476,15 @@ func (beacon *Beacon) SetThreads(threads int) {
 func IsTTDReached(chain consensus.ChainHeaderReader, parentHash common.Hash, parentNumber uint64) (bool, error) {
+	if cfg := chain.Config(); cfg.Optimism != nil {
+		num := parentNumber
+		if num == ^(uint64(0)) { // caller can (intentionally?!) underflow on parent-of-block 0 case.
```
When does this occur? When checking if IsTTDReached for the genesis block?

protolambda replied:
Yes exactly

trianglesphere commented on consensus/beacon/oplegacy.go:
```diff
@@ -0,0 +1,69 @@
+}
+
+func (o *OpLegacy) VerifyHeader(chain consensus.ChainHeaderReader, header *types.Header) error {
+	return nil // legacy chain is verified by block-hash reverse sync
```
Is this because of asking for blocks by hash? I can't find anything that explicitly checks the parent hash relationship in the headerchain / verify header code

protolambda replied:
The skeleton chain mechanism should be checking the parent-hash matches the parent block. The more-recent header that we can check the `hash(header) == recent.blockhash` of isn't here in `VerifyHeader`.

And since this is the `OpLegacy` validation, it should be used as part of the regular downloader / upon getting data from the skeleton chain. 

In the beacon variant of the consensus code there is:
```
	parent := chain.GetHeader(header.ParentHash, header.Number.Uint64()-1)
	if parent == nil {
		return consensus.ErrUnknownAncestor
	}
```
The docs of the skeleton downloader state that the headers are stored "ephemerally until sync finishes".
But since this code uses the headers DB I think it's only called in the regular downloader / after prior headers are accepted into the DB already, which means skeleton sync results are there.

So I think we may want to add this header check to ensure DB consistency, but the parent-hash checking is otherwise taken care of by the skeleton sync. Specifically the sub-chains linking/merging is what should ensure the header parent-hash consistency.

sebastianst replied:
Added this check. Tracking for `OpLegacy.VerifyHeaders` in https://github.com/ethereum-optimism/protocol-quest/issues/90

trianglesphere commented on consensus/beacon/consensus.go:
```diff
@@ -105,12 +105,27 @@ func errOut(n int, err error) chan error {
 	return errs
```
Can you add a comment on line 61 of this file that eth1 can be optimism pre-bedrock (or something to that effect)

trianglesphere reviewed (commented):
PR is looking good. It needs a some lints to be fixed, but I'd like to get this in

sebastianst reviewed (commented):
We should run a test in our snap sync infra before merging.

trianglesphere reviewed (approved):
LGTM. I've also successfully used this code to snap sync op-goerli.

trianglesphere commented on consensus/beacon/oplegacy.go:
```diff
@@ -0,0 +1,74 @@
+}
+
+func (o *OpLegacy) Finalize(chain consensus.ChainHeaderReader, header *types.Header, state *state.StateDB, txs []*types.Transaction, uncles []*types.Header, withdrawals []*types.Withdrawal) {
+	panic(fmt.Errorf("cannot finalize legacy block header: %s (num %d)", header.Hash(), header.Number))
```
We are hitting this panic in a unit test.

protolambda replied:
@trianglesphere which unit test? In the monorepo? I can help look into it

protolambda replied:
Ah I see, `TestEthClientHistoricalBackend`. It shouldn't be calling `Finalize` there. Will investigate

trianglesphere replied:
@protolambda it's getting called by CreateChain. I played around with it, but it was tricky to fix

protolambda replied:
Yep, the legacy blocks that are supposed to go into the post-bedrock node get EVM-processed upon importing, and if we use the Legacy-consensus then it's going to try to use that legacy consensus to seal the blocks.
To get around it we need to start with regular consensus, to fill our DB with the blocks (and generated receipts etc. upon importing), and then change to legacy consensus once the legacy content is prepared. I'll push a commit with a fix

protolambda commented:
Rebased on latest `optimism` branch, to resolve some conflicts with the receipts tests/code. (basefee->baseFee was renamed).

Fixed the unit-test, but had to jump through some hoops to make it work:
- historical blocks tests generates a chain
- config of above sets Bedrock at block 5
- chain import requires EVM processing
- EVM fails for < bedrock blocks, since we enforce don't support blocks to be processed pre-bedrock with EVM

Solution:
- generate with ethash faker as pre-bedrock consensus, like the upstream geth tests.
- swap the consensus of the node that we create (which by default creates a beacon consensus, wrapping an op-legacy consensus, for an optimism-configured chain-config)
- import the blocks
- swap back the consensus to op-legacy.

Related fixes:
- When loading receipts metadata, don't hydrate the L1-fee info with data if it's pre-bedrock. Don't just check if it's not a 1-tx block (as all pre-bedrock blocks are really on mainnet), since some tests have more txs than that.
- When checking if TTD has been reached, and if it's an optimism config, then enforce the bedrock block as the strict transition. Don't fall back on any TTD or ethash logic if it's not been reached yet, as there shouldn't be any other way to reach "merge" on an op-stack chain.

cc @trianglesphere, I think this unblocks this PR. PTAL
